### PR TITLE
Handle magic links for fresh sessions

### PR DESCRIPTION
**Pull request by jjcoffee**
### Details
On web we have the logic to sign in on a fresh session/different device with a "just sign in here" link, but with fresh sessions we don't have the user's permissions to the passwordless beta available, so opening magic links will always run the legacy validation (for the password-based flow) which breaks logging in for passwordless users. 

This PR modifies this behaviour so that on web, the legacy validation only runs if the user credentials have been loaded (which indicates the betas will also be available).

### Fixed Issues

$ https://github.com/Expensify/App/issues/19214
PROPOSAL: https://github.com/Expensify/App/issues/19214#issuecomment-1556965831

### Tests

On web and mobile web only (which tests the actual change):

1. Open newDot and start login process to generate a magic link email
2. Open the magic link in the email in a fresh session (e.g. in incognito mode) (make sure to change the URL from "new.expensify.com" to your local test URL)
3. Click the "sign in here" link
4. Verify that you are signed in

For native and desktop apps, you can also verify that the normal sign in flow still works:

1. Open newDot app, start sign-in
2. Enter the magic code in the app
3. Verify that you are signed in normally

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as above.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [ x The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/27287420/cbccbc0b-e79f-4d57-ab4b-498189a8217b

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/27287420/89d6dbf1-e9c9-4198-a02f-374915c2ee4e

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/27287420/b93d758d-d13f-467d-bcee-da638714c98e

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/27287420/e523958c-8e64-4af0-aeab-ce2cf34f167e

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/27287420/f3114e4a-9496-4d53-9eda-a3951bc18b64

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/27287420/cac6021f-853d-44d9-806f-0b6053be289b

</details>

### CODE REVIEW

**Comment by abdulrahuman5196** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>



</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Comment by abdulrahuman5196** [comment]
@jjcoffee  For me in native android, its still going to passing screen if I directly use the magic link without entering the email id in mobile app.
Steps:
Go to web and trigger magic link
Deeplink fresh local android build with this PR (command `npx uri-scheme open new-expensify://v/14797372/605518 --android`)
It goes to password page and not directly logging in

**Comment by jjcoffee** [comment]
@abdulrahuman5196 I am aware of that, that's why the testing steps are different for mobile - this is a web-only change (as proposed). I don't think there's a strong case for opening a magic link on mobile with a signed-out user on the app? Especially since there is no existing support for displaying the "just sign in here" link on native mobile apps (the interstitial screen doesn't display at all), so it doesn't seem to be supported behaviour. What do you think?

**Comment by jjcoffee** [comment]
@abdulrahuman5196 Just to clarify in case you want to test, if we want to adjust the native behaviour, we'd need to update the check here:

https://github.com/jjcoffee/expensify-app/blob/004d49196cc78b6e3267ab8d8cc2597bb47a4a69/src/pages/ValidateLoginPage/index.js#L46

To this: `if (!login || Permissions.canUsePasswordlessLogins(this.props.betas)) {`

We need to add the credentials to the `withOnyx` call: `credentials: {key: ONYXKEYS.CREDENTIALS},`, then we grab the login from it `const login = lodashGet(this.props, 'credentials.login', null);`.

Let me know what you think!

**Comment by abdulrahuman5196** [comment]
@jjcoffee https://github.com/Expensify/App/pull/19819#issuecomment-1571486274
Are you saying this will fix native behaviour or for testing?

**Comment by jjcoffee** [comment]
@abdulrahuman5196 Yes it will fix the native behaviour you described. I wasn't sure if you want to apply the fix since the original proposal was only for web behaviour. Shall I just push the fix do you think?

**Comment by jjcoffee** [comment]
@abdulrahuman5196 I've just pushed the fix for native behaviour, hope that's okay!

**Comment by abdulrahuman5196** [comment]
Thanks @jjcoffee, had a quick check and it seems to work. Could you update your author's checklist native platform videos on the same. I will continue review in one hour.

**Comment by abdulrahuman5196** [comment]
@jjcoffee Can you update the test to reflect the mobile native fix as well?

**Review by abdulrahuman5196 — APPROVED** [review]
Changes look's good and works well. Added Reviewers checklist as well. 

All yours
@hayata-suenaga

**Comment by abdulrahuman5196** [comment]
@hayata-suenaga  Pending on your review

**Comment by jjcoffee** [comment]
@abdulrahuman5196 Wasn't sure if it was still needed, but I've retested and updated the native videos now. Sorry for the delay!

File: src/pages/ValidateLoginPage/index.js [hunk 1]
```diff
@@ -36,14 +42,19 @@ const defaultProps = {
...
 };
 
 class ValidateLoginPage extends Component {
     componentDidMount() {
+        const login = lodashGet(this.props, 'credentials.login', null);
         const accountID = lodashGet(this.props.route.params, 'accountID', '');
         const validateCode = lodashGet(this.props.route.params, 'validateCode', '');
-        if (Permissions.canUsePasswordlessLogins(this.props.betas)) {
+
+        // A fresh session will not have credentials.login available and so also no user permission betas,
```
Comment (abdulrahuman5196) [inline src/pages/ValidateLoginPage/index.js:55]:
```suggestion
        // A fresh session will not have credentials.login and user permission betas available.
```

File: src/pages/ValidateLoginPage/index.js [hunk 2]
```diff
@@ -36,14 +42,19 @@ const defaultProps = {
...
+        // so we skip the permissions check in that case.
```
Comment (abdulrahuman5196) [inline src/pages/ValidateLoginPage/index.js:56]:
```suggestion
        // In that case, we directly allow users to go through password less flow
```

File: src/pages/ValidateLoginPage/index.website.js [hunk 3]
```diff
@@ -61,17 +61,19 @@ const defaultProps = {
 
 class ValidateLoginPage extends Component {
     componentDidMount() {
-        // Validate login if
-        // - The user is not on passwordless beta
-        if (!Permissions.canUsePasswordlessLogins(this.props.betas)) {
+        const login = lodashGet(this.props, 'credentials.login', null);
+
+        // Legacy login validation for users not on passwordless beta.
```
Comment (abdulrahuman5196) [inline src/pages/ValidateLoginPage/index.website.js:66]: Could remove this line, Not much useful here.

File: src/pages/ValidateLoginPage/index.website.js [hunk 4]
```diff
@@ -61,17 +61,19 @@ const defaultProps = {
...
+        // A fresh session will not have credentials.login available and so also no user permission betas,
```
Comment (abdulrahuman5196) [inline src/pages/ValidateLoginPage/index.website.js:67]: same suggestion as above

**Comment by abdulrahuman5196** [comment]
@jjcoffee Meanwhile could you update the comments as alone

File: src/pages/ValidateLoginPage/index.website.js (hunk 3 above)
Comment (abdulrahuman5196) [reply src/pages/ValidateLoginPage/index.website.js:66]: @jjcoffee

**Review by abdulrahuman5196 — CHANGES_REQUESTED** [review]

**Review by hayata-suenaga — DISMISSED** [review]
LGTM as long as @abdulrahuman5196's suggestions are addressed 👍

**Comment by abdulrahuman5196** [comment]
@jjcoffee Could you address the comment suggestions? We can close this out then?

File: src/pages/ValidateLoginPage/index.website.js [hunk 5]
```diff
@@ -61,17 +61,18 @@ const defaultProps = {
...
+        if (login && !Permissions.canUsePasswordlessLogins(this.props.betas)) {
             User.validateLogin(this.getAccountID(), this.getValidateCode());
             return;
         }
 
         const isSignedIn = Boolean(lodashGet(this.props, 'session.authToken', null));
         const cachedAutoAuthState = lodashGet(this.props, 'session.autoAuthState', null);
-        const login = lodashGet(this.props, 'credentials.login', null);
-        if (!login && isSignedIn && cachedAutoAuthState === CONST.AUTO_AUTH_STATE.SIGNING_IN) {
+        if (!login && isSignedIn && (cachedAutoAuthState === CONST.AUTO_AUTH_STATE.SIGNING_IN || cachedAutoAuthState === CONST.AUTO_AUTH_STATE.JUST_SIGNED_IN)) {
```
[the remark this answers is not in the document]
Comment (jjcoffee) [reply src/pages/ValidateLoginPage/index.website.js:75]: I think it's just a step missed in implementing the "just sign in here" link for signing in on a different device/session, as we do set the autoAuthState to JUST_SIGNED_IN in the `successData` when calling `signInWithValidateCode` here:

https://github.com/Expensify/App/blob/main/src/libs/actions/Session/index.js#L461-L465

**Comment by jjcoffee** [comment]
@hayata-suenaga Comments addressed!

**Review by hayata-suenaga — APPROVED** [review]

**Review by abdulrahuman5196 — APPROVED** [review]
@hayata-suenaga I think you should be good to merge

**Pull request merged**